File: frontend/src/store/audioMix.ts
// Extra audio tracks (music, voiceover, sfx) plus the dynamics chain applied to
// the mixed result. Foundation ships state, serialization and validated
// restore; the audio agent adds the mixing UI on top.

import { reactive } from 'vue'
import type {
  AudioDynamics,
  AudioRole,
  AudioTrackSpec,
  CompressorSpec,
  DuckingSpec,
  GateSpec,
  Interpolation,
  Keyframe,
  KeyframeTrack,
  LimiterSpec,
} from '../types'
import {
  assetIdOr,
  boolOr,
  clampInt,
  clampNullable,
  clampNumber,
  cloneTrack,
  enumOr,
  isRecord,
  MAX_AUDIO_TRACKS,
  MAX_KEYFRAMES,
  sanitizeKeyframeTrack,
  sanitizeList,
} from './validation'

const AUDIO_ROLES: readonly AudioRole[] = ['music', 'voiceover', 'sfx']
const MAX_TIMELINE_SECONDS = 24 * 60 * 60
const MAX_FADE_SECONDS = 60
const DEFAULT_BITRATE_KBPS = 128

/** Every dynamics field is materialised so the UI can bind to it directly. */
export interface AudioDynamicsState {
  denoise: number
  dereverb: boolean
  compressor: CompressorSpec | null
  limiter: LimiterSpec | null
  gate: GateSpec | null
  deesser: boolean
  highpassHz: number | null
  lowpassHz: number | null
  bitrateKbps: number
  volumeEnvelope: KeyframeTrack
}

export interface AudioMixState {
  tracks: AudioTrackSpec[]
  dynamics: AudioDynamicsState
}

export function defaultAudioDynamics(): AudioDynamicsState {
  return {
    denoise: 0,
    dereverb: false,
    compressor: null,
    limiter: null,
    gate: null,
    deesser: false,
    highpassHz: null,
    lowpassHz: null,
    bitrateKbps: DEFAULT_BITRATE_KBPS,
    volumeEnvelope: [],
  }
}

function defaults(): AudioMixState {
  return { tracks: [], dynamics: defaultAudioDynamics() }
}

export const audioMixState = reactive<AudioMixState>(defaults())

export function resetAudioMix(): void {
  audioMixState.tracks = []
  audioMixState.dynamics = defaultAudioDynamics()
}

function sanitizeDucking(value: unknown): DuckingSpec | null {
  if (!isRecord(value)) return null
  return {
    enabled: boolOr(value.enabled, true),
    threshold: clampNumber(value.threshold, 0, 1, 0.05),
    ratio: clampNumber(value.ratio, 1, 20, 8),
    attack: clampNumber(value.attack, 0.01, 2000, 20),
    release: clampNumber(value.release, 0.01, 9000, 300),
  }
}

/** Validate an untrusted audio track; null when it has no resolvable asset. */
export function sanitizeAudioTrack(value: unknown): AudioTrackSpec | null {
  if (!isRecord(value)) return null
  const assetId = assetIdOr(value.assetId, null)
  if (!assetId) return null
  return {
    assetId,
    role: enumOr(value.role, AUDIO_ROLES, 'music'),
    gain: clampNumber(value.gain, 0, 4, 1),
    start: clampNumber(value.start, 0, MAX_TIMELINE_SECONDS, 0),
    sourceStart: clampNumber(value.sourceStart, 0, MAX_TIMELINE_SECONDS, 0),
    end: clampNullable(value.end, 0, MAX_TIMELINE_SECONDS),
    loop: boolOr(value.loop, false),
    fadeIn: clampNumber(value.fadeIn, 0, MAX_FADE_SECONDS, 0),
    fadeOut: clampNumber(value.fadeOut, 0, MAX_FADE_SECONDS, 0),
    ducking: sanitizeDucking(value.ducking),
  }
}

function serializeAudioTrack(track: AudioTrackSpec): Record<string, unknown> {
  const out: Record<string, unknown> = {
    assetId: track.assetId,
    role: enumOr(track.role, AUDIO_ROLES, 'music'),
  }
  const gain = clampNumber(track.gain, 0, 4, 1)
  if (gain !== 1) out.gain = gain
  const start = clampNumber(track.start, 0, MAX_TIMELINE_SECONDS, 0)
  if (start !== 0) out.start = start
  const sourceStart = clampNumber(track.sourceStart, 0, MAX_TIMELINE_SECONDS, 0)
  if (sourceStart !== 0) out.sourceStart = sourceStart
  const end = clampNullable(track.end, 0, MAX_TIMELINE_SECONDS)
  if (end !== null) out.end = end
  if (track.loop) out.loop = true
  const fadeIn = clampNumber(track.fadeIn, 0, MAX_FADE_SECONDS, 0)
  if (fadeIn > 0) out.fadeIn = fadeIn
  const fadeOut = clampNumber(track.fadeOut, 0, MAX_FADE_SECONDS, 0)
  if (fadeOut > 0) out.fadeOut = fadeOut
  const ducking = sanitizeDucking(track.ducking)
  if (ducking?.enabled) out.ducking = { ...ducking }
  return out
}

function sanitizeCompressor(value: unknown): CompressorSpec | null {
  if (!isRecord(value)) return null
  return {
    threshold: clampNumber(value.threshold, -60, 0, -18),
    ratio: clampNumber(value.ratio, 1, 20, 3),
    attack: clampNumber(value.attack, 0.01, 2000, 20),
    release: clampNumber(value.release, 0.01, 9000, 250),
    makeup: clampNumber(value.makeup, 1, 64, 1),
  }
}

function sanitizeLimiter(value: unknown): LimiterSpec | null {
  if (!isRecord(value)) return null
  return { ceiling: clampNumber(value.ceiling, -30, 0, -1) }
}

function sanitizeGate(value: unknown): GateSpec | null {
  if (!isRecord(value)) return null
  return {
    threshold: clampNumber(value.threshold, -90, 0, -45),
    ratio: clampNumber(value.ratio, 1, 20, 2),
  }
}

export function sanitizeAudioDynamics(value: unknown): AudioDynamicsState {
  const source = isRecord(value) ? value : {}
  return {
    denoise: clampNumber(source.denoise, 0, 1, 0),
    dereverb: boolOr(source.dereverb, false),
    compressor: sanitizeCompressor(source.compressor),
    limiter: sanitizeLimiter(source.limiter),
    gate: sanitizeGate(source.gate),
    deesser: boolOr(source.deesser, false),
    highpassHz: clampNullable(source.highpassHz, 10, 20000),
    lowpassHz: clampNullable(source.lowpassHz, 10, 20000),
    bitrateKbps: clampInt(source.bitrateKbps, 64, 320, DEFAULT_BITRATE_KBPS),
    volumeEnvelope: sanitizeKeyframeTrack(source.volumeEnvelope, 0, 4),
  }
}

/** Only fields that differ from the neutral chain reach the wire. */
function serializeDynamics(dynamics: AudioDynamicsState): AudioDynamics {
  const out: AudioDynamics = {}
  const denoise = clampNumber(dynamics.denoise, 0, 1, 0)
  if (denoise > 0) out.denoise = denoise
  if (dynamics.dereverb) out.dereverb = true
  const compressor = sanitizeCompressor(dynamics.compressor)
  if (compressor) out.compressor = compressor
  const limiter = sanitizeLimiter(dynamics.limiter)
  if (limiter) out.limiter = limiter
  const gate = sanitizeGate(dynamics.gate)
  if (gate) out.gate = gate
  if (dynamics.deesser) out.deesser = true
  const highpassHz = clampNullable(dynamics.highpassHz, 10, 20000)
  if (highpassHz !== null) out.highpassHz = highpassHz
  const lowpassHz = clampNullable(dynamics.lowpassHz, 10, 20000)
  if (lowpassHz !== null) out.lowpassHz = lowpassHz
  const bitrateKbps = clampInt(dynamics.bitrateKbps, 64, 320, DEFAULT_BITRATE_KBPS)
  if (bitrateKbps !== DEFAULT_BITRATE_KBPS) out.bitrateKbps = bitrateKbps
  const volumeEnvelope = sanitizeKeyframeTrack(dynamics.volumeEnvelope, 0, 4)
  if (volumeEnvelope.length) out.volumeEnvelope = volumeEnvelope
  return out
}

export function audioMixPayload(): Record<string, unknown> {
  const payload: Record<string, unknown> = {}
  const tracks = sanitizeList(audioMixState.tracks, MAX_AUDIO_TRACKS, sanitizeAudioTrack)
  if (tracks.length) payload.audioTracks = tracks.map(serializeAudioTrack)
  const dynamics = serializeDynamics(audioMixState.dynamics)
  if (Object.keys(dynamics).length) payload.audioDynamics = dynamics
  return payload
}

export function applyAudioMixSnapshot(raw: unknown): void {
  const source = isRecord(raw) ? raw : {}
  audioMixState.tracks = sanitizeList(source.tracks, MAX_AUDIO_TRACKS, sanitizeAudioTrack)
  audioMixState.dynamics = sanitizeAudioDynamics(source.dynamics)
}

export function cloneAudioDynamics(dynamics: AudioDynamicsState): AudioDynamicsState {
  return {
    ...dynamics,
    compressor: dynamics.compressor ? { ...dynamics.compressor } : null,
    limiter: dynamics.limiter ? { ...dynamics.limiter } : null,
    gate: dynamics.gate ? { ...dynamics.gate } : null,
    volumeEnvelope: cloneTrack(dynamics.volumeEnvelope),
  }
}

// --- helpers for AudioMixPanel ---
//
// The panel binds to `audioMixState` directly for plain scalars; everything
// that has to stay inside the wire contract (track count, keyframe ticks,
// neutral blocks that must not leak into the payload) goes through here.

export const AUDIO_ROLE_LABELS: Record<AudioRole, string> = {
  music: 'Музыка',
  voiceover: 'Озвучка',
  sfx: 'Эффект',
}

/** Offered bitrates, all inside the 64..320 range the wire accepts. */
export const AUDIO_BITRATE_OPTIONS: readonly number[] = [64, 96, 128, 160, 192, 256, 320]

export const ENVELOPE_MIN_GAIN = 0
export const ENVELOPE_MAX_GAIN = 4

/**
 * The backend keyframe time base is 1 ms and rejects two keyframes that round
 * to the same tick, so envelope times are quantized to milliseconds here and
 * neighbours are kept at least one step apart.
 */
export const ENVELOPE_TIME_STEP = 0.001

/**
 * How the legacy `normalizeAudio` switch and the legacy volume slider interact
 * with this panel. `audio_mix::master_tail_filters` emits `loudnorm` first and
 * `volume` after it, so the slider trims the normalized level instead of being
 * swallowed by it. Keep this text in sync with that function.
 */
export const NORMALIZE_VOLUME_NOTE =
  'Нормализация приводит звук к целевому уровню -14 LUFS, а ползунок «Громкость» ' +
  'применяется после неё и сдвигает уже нормализованный уровень. Огибающая ниже ' +
  'работает раньше и меняет соотношение громких и тихих участков, а не итоговый ' +
  'уровень.'

export const BITRATE_NOTE =
  'Задаёт битрейт звуковой дорожки при экспорте. Выше битрейт: лучше качество и ' +
  'больше файл.'

export function defaultDucking(): DuckingSpec {
  return { enabled: true, threshold: 0.05, ratio: 8, attack: 20, release: 300 }
}

export function defaultCompressor(): CompressorSpec {
  return { threshold: -18, ratio: 3, attack: 20, release: 250, makeup: 1 }
}

export function defaultLimiter(): LimiterSpec {
  return { ceiling: -1 }
}

export function defaultGate(): GateSpec {
  return { threshold: -45, ratio: 2 }
}

/** A neutral extra track: full gain, at the start of the timeline, no ducking. */
export function defaultAudioTrack(assetId: string, role: AudioRole): AudioTrackSpec {
  return {
    assetId,
    role,
    gain: 1,
    start: 0,
    sourceStart: 0,
    end: null,
    loop: false,
    fadeIn: 0,
    fadeOut: 0,
    ducking: role === 'music' ? defaultDucking() : null,
  }
}

/** Append a track. Returns false for an unusable id or a full track list. */
export function addAudioTrack(assetId: unknown, role: AudioRole = 'music'): boolean {
  const id = assetIdOr(assetId, null)
  if (!id) return false
  if (audioMixState.tracks.length >= MAX_AUDIO_TRACKS) return false
  audioMixState.tracks = [
    ...audioMixState.tracks,
    defaultAudioTrack(id, enumOr(role, AUDIO_ROLES, 'music')),
  ]
  return true
}

export function removeAudioTrack(index: number): void {
  if (!Number.isInteger(index) || index < 0 || index >= audioMixState.tracks.length) return
  audioMixState.tracks = audioMixState.tracks.filter((_, position) => position !== index)
}

function quantize(seconds: number): number {
  return Math.round(seconds / ENVELOPE_TIME_STEP) * ENVELOPE_TIME_STEP
}

/** A finite, clamped, millisecond-aligned time on the output timeline. */
export function envelopeTime(value: unknown, duration: number): number {
  const max = Number.isFinite(duration) && duration > 0 ? duration : 0
  return quantize(clampNumber(value, 0, max, 0))
}

export function envelopeGain(value: unknown): number {
  return clampNumber(value, ENVELOPE_MIN_GAIN, ENVELOPE_MAX_GAIN, 1)
}

/**
 * The wire carries `interp` per keyframe, but the backend track holds one
 * interpolation for the whole track and takes it from the first keyframe. The
 * panel therefore exposes a single selector, and every point carries the same
 * value so the snapshot never claims something the render will not do.
 */
export function envelopeInterpolation(track: KeyframeTrack): Interpolation {
  return track[0]?.interp ?? 'linear'
}

export function withEnvelopeInterpolation(
  track: KeyframeTrack,
  interp: Interpolation,
): KeyframeTrack {
  return track.map((point) => ({ ...point, interp }))
}

/**
 * Insert a point, or overwrite the value when one already sits on that tick.
 * A full track (64 points) is returned unchanged.
 */
export function addEnvelopePoint(
  track: KeyframeTrack,
  seconds: number,
  value: number,
  duration: number,
): KeyframeTrack {
  const interp = envelopeInterpolation(track)
  const point: Keyframe = { t: envelopeTime(seconds, duration), v: envelopeGain(value), interp }
  const existing = track.findIndex((candidate) => candidate.t === point.t)
  if (existing >= 0) {
    return track.map((candidate, index) => (index === existing ? point : { ...candidate }))
  }
  if (track.length >= MAX_KEYFRAMES) return track
  return [...track.map((candidate) => ({ ...candidate })), point].sort((a, b) => a.t - b.t)
}

/**
 * Move one point. The time stays strictly between its neighbours so no two
 * keyframes ever collapse onto the same millisecond tick; when there is no room
 * left the point keeps its time and only the value changes.
 */
export function moveEnvelopePoint(
  track: KeyframeTrack,
  index: number,
  seconds: number,
  value: number,
  duration: number,
): KeyframeTrack {
  const current = track[index]
  if (!current) return track
  const lower = track[index - 1] ? track[index - 1]!.t + ENVELOPE_TIME_STEP : 0
  const upper = track[index + 1]
    ? track[index + 1]!.t - ENVELOPE_TIME_STEP
    : Math.max(0, Number.isFinite(duration) ? duration : 0)
  const requested = envelopeTime(seconds, duration)
  const t = lower > upper ? current.t : quantize(Math.max(lower, Math.min(upper, requested)))
  return track.map((point, position) =>
    position === index ? { ...point, t, v: envelopeGain(value) } : { ...point },
  )
}

export function removeEnvelopePoint(track: KeyframeTrack, index: number): KeyframeTrack {
  if (!track[index]) return track
  return track.filter((_, position) => position !== index).map((point) => ({ ...point }))
}

/**
 * Value of the envelope at `seconds`, mirroring the backend sampler: the track
 * holds before the first point and after the last one, and the whole track uses
 * the interpolation of its first keyframe.
 */
export function sampleEnvelope(track: KeyframeTrack, seconds: number): number {
  const first = track[0]
  if (!first) return 1
  const time = Number.isFinite(seconds) ? Math.max(0, seconds) : 0
  if (time <= first.t) return first.v
  const last = track[track.length - 1]!
  if (time >= last.t) return last.v

  let index = 1
  while (index < track.length && track[index]!.t <= time) index += 1
  const left = track[index - 1]!
  const right = track[index]!
  const span = right.t - left.t
  const progress = span > 0 ? (time - left.t) / span : 0
  let eased = progress
  if (envelopeInterpolation(track) === 'hold') {
    eased = 0
  } else if (envelopeInterpolation(track) === 'smooth') {
    eased = progress < 0.5 ? 4 * progress ** 3 : 1 - (-2 * progress + 2) ** 3 / 2
  }
  return left.v + (right.v - left.v) * eased
}

File: backend/src/domain/audio_mix.rs
//! Extra audio tracks laid onto the output timeline plus the master dynamics
//! chain (denoise, gate, compressor, de-esser, limiter, filters, envelope).

use serde::{Deserialize, Serialize};

use super::edit::{asset_reference, clamped, keyframe_track, EditSpecError};
use super::keyframes::KeyframeTrack;
use crate::model;

const MAX_GAIN: f64 = 4.0;
const MAX_TIMELINE_SECONDS: f64 = 86_400.0;
const MAX_FADE_SECONDS: f64 = 600.0;
/// Envelope times are milliseconds, matching `sidechaincompress`/`acompressor`.
const MIN_ENVELOPE_MS: f64 = 0.01;
const MAX_ENVELOPE_MS: f64 = 9_000.0;
const MIN_BITRATE_KBPS: u32 = 64;
const MAX_BITRATE_KBPS: u32 = 320;
/// Audible band limits for the optional high-pass/low-pass.
const MIN_FILTER_HZ: f64 = 10.0;
const MAX_FILTER_HZ: f64 = 20_000.0;

#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "snake_case")]
pub enum AudioRole {
    Music,
    Voiceover,
    Sfx,
}

impl AudioRole {
    pub fn parse(value: &str) -> Result<Self, EditSpecError> {
        match value {
            "music" => Ok(Self::Music),
            "voiceover" => Ok(Self::Voiceover),
            "sfx" => Ok(Self::Sfx),
            _ => Err(EditSpecError::InvalidAudioTrack),
        }
    }
}

/// Sidechain ducking of this track against the primary dialogue.
#[derive(Debug, Clone, Copy, PartialEq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase", deny_unknown_fields)]
pub struct DuckingSpec {
    pub(crate) enabled: bool,
    pub(crate) threshold: f64,
    pub(crate) ratio: f64,
    pub(crate) attack_ms: f64,
    pub(crate) release_ms: f64,
}

impl DuckingSpec {
    pub fn from_wire(value: &model::Ducking) -> Result<Self, EditSpecError> {
        Ok(Self {
            enabled: value.enabled,
            threshold: clamped(
                value.threshold,
                0.000_1,
                1.0,
                EditSpecError::InvalidAudioTrack,
            )?,
            ratio: clamped(value.ratio, 1.0, 20.0, EditSpecError::InvalidAudioTrack)?,
            attack_ms: clamped(
                value.attack,
                MIN_ENVELOPE_MS,
                MAX_ENVELOPE_MS,
                EditSpecError::InvalidAudioTrack,
            )?,
            release_ms: clamped(
                value.release,
                MIN_ENVELOPE_MS,
                MAX_ENVELOPE_MS,
                EditSpecError::InvalidAudioTrack,
            )?,
        })
    }

    pub fn enabled(self) -> bool {
        self.enabled
    }

    pub fn threshold(self) -> f64 {
        self.threshold
    }

    pub fn ratio(self) -> f64 {
        self.ratio
    }

    pub fn attack_ms(self) -> f64 {
        self.attack_ms
    }

    pub fn release_ms(self) -> f64 {
        self.release_ms
    }
}

#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase", deny_unknown_fields)]
pub struct AudioTrackSpec {
    pub(crate) asset_id: String,
    pub(crate) role: AudioRole,
    pub(crate) gain: f64,
    /// Where the track lands on the OUTPUT timeline.
    pub(crate) start_seconds: f64,
    /// In-point inside the asset.
    pub(crate) source_start_seconds: f64,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub(crate) end_seconds: Option<f64>,
    pub(crate) looping: bool,
    pub(crate) fade_in_seconds: f64,
    pub(crate) fade_out_seconds: f64,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub(crate) ducking: Option<DuckingSpec>,
}

impl AudioTrackSpec {
    pub fn from_wire(value: &model::AudioTrack) -> Result<Self, EditSpecError> {
        let spec = Self {
            asset_id: asset_reference(&value.asset_id)?,
            role: AudioRole::parse(&value.role)?,
            gain: clamped(value.gain, 0.0, MAX_GAIN, EditSpecError::InvalidAudioTrack)?,
            start_seconds: clamped(
                value.start,
                0.0,
                MAX_TIMELINE_SECONDS,
                EditSpecError::InvalidAudioTrack,
            )?,
            source_start_seconds: clamped(
                value.source_start,
                0.0,
                MAX_TIMELINE_SECONDS,
                EditSpecError::InvalidAudioTrack,
            )?,
            end_seconds: value
                .end
                .map(|end| {
                    clamped(
                        end,
                        0.0,
                        MAX_TIMELINE_SECONDS,
                        EditSpecError::InvalidAudioTrack,
                    )
                })
                .transpose()?,
            looping: value.r#loop,
            fade_in_seconds: clamped(
                value.fade_in,
                0.0,
                MAX_FADE_SECONDS,
                EditSpecError::InvalidAudioTrack,
            )?,
            fade_out_seconds: clamped(
                value.fade_out,
                0.0,
                MAX_FADE_SECONDS,
                EditSpecError::InvalidAudioTrack,
            )?,
            ducking: value
                .ducking
                .as_ref()
                .map(DuckingSpec::from_wire)
                .transpose()?,
        };
        spec.validate()?;
        Ok(spec)
    }

    pub fn asset_id(&self) -> &str {
        &self.asset_id
    }

    pub fn role(&self) -> AudioRole {
        self.role
    }

    pub fn gain(&self) -> f64 {
        self.gain
    }

    pub fn start_seconds(&self) -> f64 {
        self.start_seconds
    }

    pub fn source_start_seconds(&self) -> f64 {
        self.source_start_seconds
    }

    pub fn end_seconds(&self) -> Option<f64> {
        self.end_seconds
    }

    pub fn looping(&self) -> bool {
        self.looping
    }

    pub fn fade_in_seconds(&self) -> f64 {
        self.fade_in_seconds
    }

    pub fn fade_out_seconds(&self) -> f64 {
        self.fade_out_seconds
    }

    pub fn ducking(&self) -> Option<DuckingSpec> {
        self.ducking
    }

    pub(crate) fn validate(&self) -> Result<(), EditSpecError> {
        if self.asset_id.is_empty()
            || !(0.0..=MAX_GAIN).contains(&self.gain)
            || self.start_seconds < 0.0
            || self.source_start_seconds < 0.0
        {
            return Err(EditSpecError::InvalidAudioTrack);
        }
        if let Some(end) = self.end_seconds {
            if end <= self.start_seconds {
                return Err(EditSpecError::InvalidAudioTrack);
            }
        }
        Ok(())
    }
}

#[derive(Debug, Clone, Copy, PartialEq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase", deny_unknown_fields)]
pub struct CompressorSpec {
    pub(crate) threshold_db: f64,
    pub(crate) ratio: f64,
    pub(crate) attack_ms: f64,
    pub(crate) release_ms: f64,
    pub(crate) makeup: f64,
}

impl CompressorSpec {
    pub fn from_wire(value: &model::Compressor) -> Result<Self, EditSpecError> {
        Ok(Self {
            threshold_db: clamped(
                value.threshold,
                -60.0,
                0.0,
                EditSpecError::InvalidAudioDynamics,
            )?,
            ratio: clamped(value.ratio, 1.0, 20.0, EditSpecError::InvalidAudioDynamics)?,
            attack_ms: clamped(
                value.attack,
                MIN_ENVELOPE_MS,
                MAX_ENVELOPE_MS,
                EditSpecError::InvalidAudioDynamics,
            )?,
            release_ms: clamped(
                value.release,
                MIN_ENVELOPE_MS,
                MAX_ENVELOPE_MS,
                EditSpecError::InvalidAudioDynamics,
            )?,
            makeup: clamped(value.makeup, 1.0, 64.0, EditSpecError::InvalidAudioDynamics)?,
        })
    }

    pub fn threshold_db(self) -> f64 {
        self.threshold_db
    }

    pub fn ratio(self) -> f64 {
        self.ratio
    }

    pub fn attack_ms(self) -> f64 {
        self.attack_ms
    }

    pub fn release_ms(self) -> f64 {
        self.release_ms
    }

    pub fn makeup(self) -> f64 {
        self.makeup
    }
}

#[derive(Debug, Clone, Copy, PartialEq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase", deny_unknown_fields)]
pub struct LimiterSpec {
    pub(crate) ceiling_db: f64,
}

impl LimiterSpec {
    pub fn from_wire(value: &model::Limiter) -> Result<Self, EditSpecError> {
        Ok(Self {
            ceiling_db: clamped(
                value.ceiling,
                -30.0,
                0.0,
                EditSpecError::InvalidAudioDynamics,
            )?,
        })
    }

    pub fn ceiling_db(self) -> f64 {
        self.ceiling_db
    }
}

#[derive(Debug, Clone, Copy, PartialEq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase", deny_unknown_fields)]
pub struct GateSpec {
    pub(crate) threshold_db: f64,
    pub(crate) ratio: f64,
}

impl GateSpec {
    pub fn from_wire(value: &model::Gate) -> Result<Self, EditSpecError> {
        Ok(Self {
            threshold_db: clamped(
                value.threshold,
                -80.0,
                0.0,
                EditSpecError::InvalidAudioDynamics,
            )?,
            ratio: clamped(value.ratio, 1.0, 20.0, EditSpecError::InvalidAudioDynamics)?,
        })
    }

    pub fn threshold_db(self) -> f64 {
        self.threshold_db
    }

    pub fn ratio(self) -> f64 {
        self.ratio
    }
}

#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase", deny_unknown_fields)]
pub struct AudioDynamicsSpec {
    pub(crate) denoise: f64,
    pub(crate) dereverb: bool,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub(crate) compressor: Option<CompressorSpec>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub(crate) limiter: Option<LimiterSpec>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub(crate) gate: Option<GateSpec>,
    pub(crate) deesser: bool,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub(crate) highpass_hz: Option<f64>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub(crate) lowpass_hz: Option<f64>,
    pub(crate) bitrate_kbps: u32,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub(crate) volume_envelope: Option<KeyframeTrack<f64>>,
}

impl AudioDynamicsSpec {
    pub fn from_wire(value: &model::AudioDynamics) -> Result<Self, EditSpecError> {
        let spec = Self {
            denoise: clamped(value.denoise, 0.0, 1.0, EditSpecError::InvalidAudioDynamics)?,
            dereverb: value.dereverb,
            compressor: value
                .compressor
                .as_ref()
                .map(CompressorSpec::from_wire)
                .transpose()?,
            limiter: value
                .limiter
                .as_ref()
                .map(LimiterSpec::from_wire)
                .transpose()?,
            gate: value.gate.as_ref().map(GateSpec::from_wire).transpose()?,
            deesser: value.deesser,
            highpass_hz: value
                .highpass_hz
                .map(|hz| {
                    clamped(
                        hz,
                        MIN_FILTER_HZ,
                        MAX_FILTER_HZ,
                        EditSpecError::InvalidAudioDynamics,
                    )
                })
                .transpose()?,
            lowpass_hz: value
                .lowpass_hz
                .map(|hz| {
                    clamped(
                        hz,
                        MIN_FILTER_HZ,
                        MAX_FILTER_HZ,
                        EditSpecError::InvalidAudioDynamics,
                    )
                })
                .transpose()?,
            bitrate_kbps: value.bitrate_kbps.clamp(MIN_BITRATE_KBPS, MAX_BITRATE_KBPS),
            volume_envelope: keyframe_track(
                &value.volume_envelope,
                EditSpecError::InvalidAudioDynamics,
            )?,
        };
        spec.validate()?;
        Ok(spec)
    }

    pub fn denoise(&self) -> f64 {
        self.denoise
    }

    pub fn dereverb(&self) -> bool {
        self.dereverb
    }

    pub fn compressor(&self) -> Option<CompressorSpec> {
        self.compressor
    }

    pub fn limiter(&self) -> Option<LimiterSpec> {
        self.limiter
    }

    pub fn gate(&self) -> Option<GateSpec> {
        self.gate
    }

    pub fn deesser(&self) -> bool {
        self.deesser
    }

    pub fn highpass_hz(&self) -> Option<f64> {
        self.highpass_hz
    }

    pub fn lowpass_hz(&self) -> Option<f64> {
        self.lowpass_hz
    }

    pub fn bitrate_kbps(&self) -> u32 {
        self.bitrate_kbps
    }

    pub fn volume_envelope(&self) -> Option<&KeyframeTrack<f64>> {
        self.volume_envelope.as_ref()
    }

    pub(crate) fn validate(&self) -> Result<(), EditSpecError> {
        if !(0.0..=1.0).contains(&self.denoise)
            || !(MIN_BITRATE_KBPS..=MAX_BITRATE_KBPS).contains(&self.bitrate_kbps)
        {
            return Err(EditSpecError::InvalidAudioDynamics);
        }
        if let (Some(high), Some(low)) = (self.highpass_hz, self.lowpass_hz) {
            if high >= low {
                return Err(EditSpecError::InvalidAudioDynamics);
            }
        }
        Ok(())
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    fn track() -> model::AudioTrack {
        model::AudioTrack {
            asset_id: "ast_0123456789abcdef".into(),
            role: "music".into(),
            gain: 1.0,
            start: 0.0,
            source_start: 0.0,
            end: None,
            r#loop: false,
            fade_in: 0.0,
            fade_out: 0.0,
            ducking: None,
        }
    }

    fn dynamics() -> model::AudioDynamics {
        model::AudioDynamics {
            denoise: 0.0,
            dereverb: false,
            compressor: None,
            limiter: None,
            gate: None,
            deesser: false,
            highpass_hz: None,
            lowpass_hz: None,
            bitrate_kbps: 128,
            volume_envelope: Vec::new(),
        }
    }

    #[test]
    fn track_gain_and_role_are_validated() {
        let mut loud = track();
        loud.gain = 99.0;
        assert_eq!(AudioTrackSpec::from_wire(&loud).unwrap().gain(), MAX_GAIN);

        let mut unknown = track();
        unknown.role = "narration".into();
        assert_eq!(
            AudioTrackSpec::from_wire(&unknown),
            Err(EditSpecError::InvalidAudioTrack)
        );

        let mut inverted = track();
        inverted.start = 10.0;
        inverted.end = Some(1.0);
        assert_eq!(
            AudioTrackSpec::from_wire(&inverted),
            Err(EditSpecError::InvalidAudioTrack)
        );
    }

    #[test]
    fn non_finite_values_never_reach_the_adapter() {
        let mut broken = track();
        broken.gain = f64::NAN;
        assert_eq!(
            AudioTrackSpec::from_wire(&broken),
            Err(EditSpecError::InvalidAudioTrack)
        );

        let mut ducked = track();
        ducked.ducking = Some(model::Ducking {
            enabled: true,
            threshold: f64::INFINITY,
            ratio: 8.0,
            attack: 20.0,
            release: 300.0,
        });
        assert_eq!(
            AudioTrackSpec::from_wire(&ducked),
            Err(EditSpecError::InvalidAudioTrack)
        );
    }

    #[test]
    fn dynamics_clamp_the_bitrate_and_reject_crossed_filters() {
        let mut low = dynamics();
        low.bitrate_kbps = 8;
        assert_eq!(
            AudioDynamicsSpec::from_wire(&low).unwrap().bitrate_kbps(),
            MIN_BITRATE_KBPS
        );

        let mut crossed = dynamics();
        crossed.highpass_hz = Some(8_000.0);
        crossed.lowpass_hz = Some(200.0);
        assert_eq!(
            AudioDynamicsSpec::from_wire(&crossed),
            Err(EditSpecError::InvalidAudioDynamics)
        );
    }

    #[test]
    fn volume_envelope_becomes_a_millisecond_keyframe_track() {
        let mut enveloped = dynamics();
        enveloped.volume_envelope = vec![
            model::Keyframe {
                t: 0.0,
                v: 1.0,
                interp: model::KeyframeInterpolation::Linear,
            },
            model::Keyframe {
                t: 2.5,
                v: 0.2,
                interp: model::KeyframeInterpolation::Linear,
            },
        ];
        let spec = AudioDynamicsSpec::from_wire(&enveloped).unwrap();
        let envelope = spec.volume_envelope().unwrap();
        assert_eq!(envelope.keyframes[1].tick, 2_500);
        assert_eq!(envelope.sample_seconds(2.5).unwrap(), 0.2);
    }
}
